### ISSUE TITLE
Symfony/Doctrine fail to find an entity based off of its binary UUID

### ISSUE THREAD

**Issue by BenoitDuffez** [ramsey/uuid-doctrine#264]
I have created a new Symfony app with version 7.0.7 and doctrine/orm ^3.1. I followed the instructions to use binary UUIDs and then created a CRUD for my entities. Entity A is standalone and works well. Entity B has a ManyToOne relation to A and while the form displays fine, I can't validate it and thus create a new B.

I was able to pin the issue here, in `vendor/symfony/doctrine-bridge/Form/ChoiceList/ORMQueryBuilderLoader`, line 70:

```php
            // Filter out non-integer values (e.g. ""). If we don't, some
            // databases such as PostgreSQL fail.
            $values = array_values(array_filter($values, fn ($v) => (string) $v === (string) (int) $v || ctype_digit($v)));
        } elseif (\in_array($type, ['ulid', 'uuid', 'guid'])) { // <-----
            $parameterType = ArrayParameterType::STRING;
```

now if I modify this to:

```php
            // Filter out non-integer values (e.g. ""). If we don't, some
            // databases such as PostgreSQL fail.
            $values = array_values(array_filter($values, fn ($v) => (string) $v === (string) (int) $v || ctype_digit($v)));
        } elseif (\in_array($type, ['ulid', 'uuid', 'guid', 'uuid_binary'])) {
            $parameterType = ArrayParameterType::STRING;
```

then it works like a charm.

I feel like I have missed something in the setup of this lib but I can't pinpoint what/where.

`config/packages/doctrine.yaml`

```yaml
doctrine:
    dbal:
        url: '%env(resolve:DATABASE_URL)%'
        profiling_collect_backtrace: '%kernel.debug%'
        use_savepoints: true
        types:
            uuid_binary:  Ramsey\Uuid\Doctrine\UuidBinaryType
```

I also tried this in my Kernel but that does not work either:

```php
public function boot(): void {
     parent::boot();
    \Doctrine\DBAL\Types\Type::addType('uuid_binary', 'Ramsey\Uuid\Doctrine\UuidBinaryType');
    $entityManager = $this->getContainer()->get('doctrine.orm.default_entity_manager');
    $entityManager->getConnection()->getDatabasePlatform()->registerDoctrineTypeMapping('uuid_binary', 'binary');
}
```

**Comment by subdee** [comment]
In your case, binding a string parameter won't work, because the column is not a CHAR type but a BINARY type. 

A BINARY type though does not seem to be supported in doctrine.

If I bind the binary representation of the ID in a param I get an error from the ORM.

This is the parameter:
```php
->setParameter(':column', Uuid::fromString($client->getId()->toString())->getBytes())
```
and the error is `Unhandled match case 16`  which is coming from [this method in doctrine/orm](https://github.com/doctrine/orm/blob/3.1.x/src/Persisters/Entity/BasicEntityPersister.php#L1908) and is simply missing a match for `ParameterType::BINARY`.

Adding the type explicitly results in the same, binding it as string has the same effect as @BenoitDuffez, where no results are returned.

**Comment by BenoitDuffez** [comment]
Your comment is valid for querying the database. Indeed we need to set a parameter to the bytes of the UUID. One note about your code though, I think it would be easier with

```php
->setParameter(':column', $client->getId()->getBytes())
```

indeed, the `getId` returns a UUID that you can extract the bytes from.

However, my ticket was about forms. If you declare relationships between two entities, they are linked with their UUID and it works fine in the database. The problem is that when saving the form, the form data type is `uuid_binary` and Symfony has a static check that does not include this (but does include `uuid`).

Changing Symfony code is wrong, perhaps an upstream patch would be accepted but it also feels wrong to me. I don't know if there are events or configuration items that can be leveraged by this library so that Symfony knows how to handle `uuid_binary`.

PS: by "Symfony" I mean `symfony/doctrine-bridge`

**Comment by subdee** [comment]
@BenoitDuffez You are right, they are not related, I will make a separate issue for my problem.

**Comment by pluk77** [comment]
Is this still an issue?

Have you tried adding an alias in your ramsey_uuid_doctrine.yaml file so you can use uuid in your annotation or attributes of your entities?

```
doctrine:
    dbal:
        types:
            uuid: Ramsey\Uuid\Doctrine\UuidBinaryType
            uuid_binary: Ramsey\Uuid\Doctrine\UuidBinaryType
```

**Comment by BenoitDuffez** [comment]
I think this is essentially what I had tried in the code in my OP?

```php
public function boot(): void {
     parent::boot();
    \Doctrine\DBAL\Types\Type::addType('uuid_binary', 'Ramsey\Uuid\Doctrine\UuidBinaryType');
    $entityManager = $this->getContainer()->get('doctrine.orm.default_entity_manager');
    $entityManager->getConnection()->getDatabasePlatform()->registerDoctrineTypeMapping('uuid_binary', 'binary');
}
```

If not I'll try to see if I reproduce.

**Comment by pluk77** [comment]
I am not 100% sure if it your OP code has the same effect. never done it that way.

We have been using binary uuids for a long time and I have not come across the issue you reported in recent times. 

What worked for us is to register both uuid and uuid_binary for the UuidBinaryType AND use 'uuid' and not 'uuid_binary' in our entity ORM annotations and/or attributed.

**Comment by BenoitDuffez** [comment]
I'll be honest, the code I used last year that triggered that issue is no longer in prod and it would take a significant amount of time to put it back together for testing. I suppose we can close this ticket and I'll reopen if I reproduce.
